pkg/farm/farm.go

```diff
@@ -0,0 +1,503 @@
+	var localPlatform string
+	// Make notes of which platforms we can build for natively, and which
+	// ones we can build for using emulation.
+	for name, engine := range f.builders {
+		name, engine := name, engine
+		infoGroup.Go(func() error {
+			inspect, err := engine.FarmNodeInspect(ctx)
+			if err != nil {
+				return err
+			}
+			infoMutex.Lock()
+			defer infoMutex.Unlock()
+			for _, n := range inspect.NativePlatforms {
+				if _, assigned := native[n]; !assigned {
+					native[n] = name
+				}
+				if name == entities.LocalFarmImageBuilderName {
+					localPlatform = n
+				}
+			}
+			for _, e := range inspect.EmulatedPlatforms {
+				if _, assigned := emulated[e]; !assigned {
+					emulated[e] = name
+				}
+			}
+			return nil
+		})
+	}
+	merr := infoGroup.Wait()
+	if merr != nil {
+		if err := merr.ErrorOrNil(); err != nil {
+			return Schedule{}, err
+		}
+	}
+	// Assign a build to the first node that could build it natively, and
+	// if there isn't one, the first one that can build it with the help of
+	// emulation, and if there aren't any, error out.
+	for _, platform := range platforms {
+		if builder, ok := native[platform]; ok {
+			platformBuilders[platform] = builder
+		} else if builder, ok := emulated[platform]; ok {
+			platformBuilders[platform] = builder
+		} else {
+			return Schedule{}, fmt.Errorf("no builder capable of building for platform %q available", platform)
+		}
+	}
+	// If local is set, prioritize building on local
+	if localPlatform != "" {
+		platformBuilders[localPlatform] = entities.LocalFarmImageBuilderName
+	}
+	schedule := Schedule{
+		platformBuilders: platformBuilders,
+	}
+	return schedule, nil
+}
+
+// Build runs a build using the specified targetplatform:service map.  If all
+// builds succeed, it copies the resulting images from the remote hosts to the
+// local service and builds a manifest list with the specified reference name.
+func (f *Farm) Build(ctx context.Context, schedule Schedule, options entities.BuildOptions, reference string) error {
+	switch options.OutputFormat {
+	default:
+		return fmt.Errorf("unknown output format %q requested", options.OutputFormat)
+	case "", define.OCIv1ImageManifest:
+		options.OutputFormat = define.OCIv1ImageManifest
+	case define.Dockerv2ImageManifest:
+	}
+
+	// Build the list of jobs.
+	var jobs sync.Map
+	type job struct {
+		platform string
+		os       string
+		arch     string
+		variant  string
+		builder  entities.ImageEngine
+	}
+	for platform, builderName := range schedule.platformBuilders { // prepare to build
+		builder, ok := f.builders[builderName]
+		if !ok {
+			return fmt.Errorf("unknown builder %q", builderName)
+		}
+		var rawOS, rawArch, rawVariant string
+		p := strings.Split(platform, "/")
+		if len(p) > 0 && p[0] != "" {
+			rawOS = p[0]
+		}
+		if len(p) > 1 {
+			rawArch = p[1]
+		}
+		if len(p) > 2 {
+			rawVariant = p[2]
+		}
+		os, arch, variant := libimage.NormalizePlatform(rawOS, rawArch, rawVariant)
```

Comment:
you need to split out NormalizePlatform() from libimage into another package to avoid the full import of libimage if you have to use this function.

Reply:
fixed